x/bank/keeper/deferred_cache.go

```diff
@@ -0,0 +1,126 @@
+package keeper
+
+import (
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/bank/types"
+)
+
+type DeferredCache struct {
+	storeKey sdk.StoreKey
+	cdc      codec.BinaryCodec
+}
+
+func NewDeferredCache(cdc codec.BinaryCodec, storeKey sdk.StoreKey) *DeferredCache {
+	return &DeferredCache{
+		cdc:      cdc,
+		storeKey: storeKey,
+	}
+}
+
+func (d *DeferredCache) getModuleTxIndexedStore(ctx sdk.Context, moduleAddr sdk.AccAddress, txIndex uint64) prefix.Store {
+	store := ctx.KVStore(d.storeKey)
+
+	return prefix.NewStore(store, types.CreateDeferredCacheModuleTxIndexedPrefix(moduleAddr, txIndex))
+}
+
+// GetBalance returns the balance of a specific denomination for a given module address and transaction index
+func (d *DeferredCache) GetBalance(ctx sdk.Context, moduleAddr sdk.AccAddress, txIndex uint64, denom string) sdk.Coin {
+	deferredStore := d.getModuleTxIndexedStore(ctx, moduleAddr, txIndex)
+
+	bz := deferredStore.Get([]byte(denom))
+	if bz == nil {
+		return sdk.NewCoin(denom, sdk.ZeroInt())
+	}
+
+	var balance sdk.Coin
+	d.cdc.MustUnmarshal(bz, &balance)
+
+	return balance
+}
+
+// setBalance sets the coin balance for a module and tx Index.
+func (d *DeferredCache) setBalance(ctx sdk.Context, moduleAddr sdk.AccAddress, txIndex uint64, balance sdk.Coin) error {
+	if !balance.IsValid() {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidCoins, balance.String())
+	}
+
+	deferredStore := d.getModuleTxIndexedStore(ctx, moduleAddr, txIndex)
+	// Bank invariants require to not store zero balances, so we follow the same pattern in deferred cache.
+	if balance.IsZero() {
```

Comment:
nit but could we add unit tests for failure scenarios? The happy paths seem to be covered quite well already. Not a blocker for this PR